### PR TITLE
LinearBVH cleanup

### PR DESCRIPTION
**Pull request by kanye-quest**
# Summary

- Removes `linear_bvh::AABB`, `linear_bvh::Vec`, and `linear_bvh::Range`; and replaces them with `primal::BoundingBox` and `primal::Point/Vector`.
- Make some more functions on `primal::BoundingBox` and `primal::Vector` host/device-enabled.
- Enable using `spin::Mortonizer` and `convertPointToMorton()` functions on the GPU.
- Use implicitly-generated copy and assignment constructors for `BoundingBox`,`Vector`, `Point`, and `NumericArray`; this seems to improve performance of the BVH traversal on the GPU.

### CODE REVIEW

File: src/axom/spin/internal/linear_bvh/emit_bvh.hpp [hunk 1]
```diff
@@ -106,21 +106,21 @@ void emit_bvh(RadixTree<FloatType, 3>& data, BVHData<FloatType, 3>& bvh_data)
                                // do the offset now
                                rchild *= 4;
                              }
-                             vec1[0] = l_aabb.m_x.min();
-                             vec1[1] = l_aabb.m_y.min();
-                             vec1[2] = l_aabb.m_z.min();
+                             vec1[0] = l_aabb.getMin()[0];
```
Comment (kennyweiss) [inline src/axom/spin/internal/linear_bvh/emit_bvh.hpp:109]: Could this use something like:
```cpp
primal::Vector<FloatType, 4> vec1(l_aabb.getMin().array());
```
or
```cpp
primal::Vector<FloatType, 4> vec1 = l_aabb.getMin().array();
```
(and similarly for the others)

Or is it better to explicitly set each entry within the kernel for some reason?

File: src/axom/spin/tests/spin_bvh.cpp [hunk 2]
```diff
@@ -318,18 +309,19 @@ void generate_aabbs_and_centroids2d(const mint::Mesh* mesh,
         xsum += node[mint::X_COORDINATE];
         ysum += node[mint::Y_COORDINATE];
 
-        xrange.include(node[mint::X_COORDINATE]);
-        yrange.include(node[mint::Y_COORDINATE]);
+        range.addPoint(
+          primal::Point<FloatType, 2>::make_point(node[mint::X_COORDINATE],
```
Comment (kennyweiss) [inline src/axom/spin/tests/spin_bvh.cpp:313]: Minor: Would the following work?
```cpp
        range.addPoint( primal::Point<FloatType, 2>{ node[mint::X_COORDINATE], 
                                                     node[mint::Y_COORDINATE] });
```

**Review by kennyweiss — APPROVED** [review]
This is great -- thanks @kanye-quest !

File: src/axom/spin/tests/spin_bvh.cpp (hunk 2 above)
Comment (kanye-quest) [reply src/axom/spin/tests/spin_bvh.cpp:313]: The issue is that `node` is an array of doubles, but `FloatType` can be instantiated as either a float or a double. So the initializer list overload doesn't match.

Comment (kennyweiss) [reply src/axom/spin/tests/spin_bvh.cpp:313]: That makes sense -- thanks.

Alternatively, I suppose you can static cast, which might make that distinction a bit more explicit.

File: src/axom/spin/internal/linear_bvh/emit_bvh.hpp (hunk 1 above)
Comment (kanye-quest) [reply src/axom/spin/internal/linear_bvh/emit_bvh.hpp:109]: The code here packs the 3D min/max points from two bounding boxes into 4-wide vectors. We could do a direct assignment for `l_aabb.getMin() -> vec1`, but the other points are split up between two of the vec4.

Comment (kennyweiss) [reply src/axom/spin/internal/linear_bvh/emit_bvh.hpp:109]: Thanks.  You're right, I wasn't paying attention to the indexing.

File: src/axom/spin/tests/spin_bvh.cpp (hunk 2 above)
Comment (kanye-quest) [reply src/axom/spin/tests/spin_bvh.cpp]: I'll go with initializer lists + static_cast; it's probably best to be explicit wrt the floating-point conversion.

**Review by bmhan12 — APPROVED** [review]
LGTM!

**Comment by rhornung67** [comment]
@bmhan12 the "LGTM" comment is no longer needed to launch the Gitlab CI, or were you just making a comment?

**Review by rhornung67 — APPROVED** [review]
One nit-pick comment. Otherwise, looks great!

**Review by agcapps — APPROVED** [review]
Max, thanks for your work on this.

**Comment by bmhan12** [comment]
> @bmhan12 the "LGTM" comment is no longer needed to launch the Gitlab CI, or were you just making a comment?

Just the comment 👍

File: src/axom/primal/geometry/BoundingBox.hpp [hunk 3]
```diff
@@ -511,7 +511,7 @@ template <typename T, int NDIMS>
 BoundingBox<T, NDIMS>& BoundingBox<T, NDIMS>::scale(double scaleFactor)
 {
   const PointType midpoint = getCentroid();
-  const VectorType r = (T)(scaleFactor * 0.5) * range();
+  const VectorType r = static_cast<T>(scaleFactor * 0.5) * range();
```
Comment (rhornung67) [inline src/axom/primal/geometry/BoundingBox.hpp:514]: 👍

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**